pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -0,0 +1,279 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+/// Edit this file to define custom logic or remove it if it is not needed.
+/// Learn more about FRAME and the core library of Substrate FRAME pallets:
+/// <https://docs.substrate.io/v3/runtime/frame>
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+mod types;
+
+#[frame_support::pallet(dev_mode)]
+pub mod pallet {
+	use crate::types::{Curve, MockCurve, PoolDetails, PoolStatus, TokenMeta};
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::*,
+		traits::{
+			fungible::{Inspect as InspectFungible, Mutate, MutateHold},
+			fungibles::{
+				metadata::Mutate as FungiblesMetadata, Create as CreateFungibles, Destroy as DestroyFungibles,
+				Inspect as InspectFungibles, Mutate as MutateFungibles,
+			},
+			tokens::Preservation,
+		},
+		Hashable,
+	};
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::{
+		traits::{CheckedAdd, CheckedConversion, Saturating, StaticLookup, Zero},
+		ArithmeticError, SaturatedConversion,
+	};
+
+	type AccountIdLookupOf<T> = <<T as frame_system::Config>::Lookup as sp_runtime::traits::StaticLookup>::Source;
+	pub type DepositCurrencyBalanceOf<T> =
+		<<T as Config>::DepositCurrency as InspectFungible<<T as frame_system::Config>::AccountId>>::Balance;
+	pub type DepositCurrencyHoldReasonOf<T> =
+		<<T as Config>::DepositCurrency as frame_support::traits::fungible::InspectHold<
+			<T as frame_system::Config>::AccountId,
+		>>::Reason;
+	pub type CollateralCurrencyBalanceOf<T> =
+		<<T as Config>::CollateralCurrency as InspectFungible<<T as frame_system::Config>::AccountId>>::Balance;
+	pub type FungiblesBalanceOf<T> =
+		<<T as Config>::Fungibles as InspectFungibles<<T as frame_system::Config>::AccountId>>::Balance;
+	pub type FungiblesAssetIdOf<T> =
+		<<T as Config>::Fungibles as InspectFungibles<<T as frame_system::Config>::AccountId>>::AssetId;
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The currency used for storage deposits.
+		type DepositCurrency: MutateHold<Self::AccountId>;
+		/// The currency used as collateral for minting bonded tokens.
+		type CollateralCurrency: Mutate<Self::AccountId>;
+		/// Implementation of creating and managing new fungibles
+		type Fungibles: CreateFungibles<Self::AccountId>
+			+ DestroyFungibles<Self::AccountId>
+			+ FungiblesMetadata<Self::AccountId>
+			+ MutateFungibles<Self::AccountId>;
+		/// The maximum number of currencies allowed for a single pool.
+		#[pallet::constant]
+		type MaxCurrencies: Get<u32> + TypeInfo;
```

Comment:
This is benchmarking/weight relevant though, won't this be a problem?

Reply:
I don't think this will be an issue for benchmarking. We can't anticipate the worst-case scenario for u32—that would be too much. I suggest moving the currency insertion logic into a separate function and benchmarking it in isolation with one currency. The final weight can then be calculated by multiplying the benchmarked weight by the number of currencies.